File: ios/FocusDhikr/Support/Theme.swift
import SwiftUI

/// A quiet palette.
///
/// Warm near-black and parchment rather than pure black and white, one muted
/// gold accent, and nothing else. No progress bars racing to full, no confetti,
/// no streak flames. Closer to a well-set book than to a game.
enum FocusColor {
    static let ink = Color(red: 0.07, green: 0.06, blue: 0.05)
    static let inkElevated = Color(red: 0.11, green: 0.10, blue: 0.09)
    static let textOnInk = Color(red: 0.91, green: 0.89, blue: 0.85)
    static let textMuted = Color(red: 0.66, green: 0.63, blue: 0.60)
    static let outline = Color(red: 0.16, green: 0.15, blue: 0.14)
    /// The single accent. Used sparingly: remaining time, the active step, dhikr.
    static let gold = Color(red: 0.79, green: 0.64, blue: 0.15)
    /// Never red-for-shame. This is the "you reached your limit" tone.
    static let clay = Color(red: 0.69, green: 0.44, blue: 0.34)
}

enum Spacing {
    static let xs: CGFloat = 4
    static let sm: CGFloat = 8
    static let md: CGFloat = 16
    static let lg: CGFloat = 24
    static let xl: CGFloat = 32
    static let xxl: CGFloat = 48
}

/// Serif for anything meant to be read slowly, sans for anything scanned.
/// Both are system faces, so the app ships no font binaries.
extension Font {
    static let focusDisplay = Font.system(size: 32, weight: .regular, design: .serif)
    static let focusHeadline = Font.system(size: 24, weight: .regular, design: .serif)
    static let focusTitle = Font.system(size: 17, weight: .medium)
    static let focusBody = Font.system(size: 16)
    static let focusCaption = Font.system(size: 12, weight: .medium)
}

/// The only container shape in the app. One radius, one border, no shadows.
struct QuietCard<Content: View>: View {
    @ViewBuilder var content: Content

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.sm) {
            content
        }
        .padding(Spacing.md)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(FocusColor.inkElevated)
        .clipShape(RoundedRectangle(cornerRadius: 18, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: 18, style: .continuous)
                .stroke(FocusColor.outline, lineWidth: 1)
        )
    }
}

struct PrimaryAction: View {
    let title: String
    var enabled: Bool = true
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(title)
                .font(.focusTitle)
                .frame(maxWidth: .infinity)
                .padding(.vertical, 16)
        }
        .disabled(!enabled)
        .background(enabled ? FocusColor.textOnInk : FocusColor.outline)
        .foregroundStyle(enabled ? FocusColor.ink : FocusColor.textMuted)
        .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
    }
}

/// The "turn back" affordance.
///
/// Deliberately given equal visual weight to continuing, never made harder to
/// find or smaller. The app does not trick the user in either direction.
struct QuietAction: View {
    let title: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(title)
                .font(.focusTitle)
                .foregroundStyle(FocusColor.textMuted)
                .frame(maxWidth: .infinity)
                .padding(.vertical, 14)
        }
    }
}

/// Dots marking progress through the gate.
///
/// Not a progress bar: a bar racing to 100% invites you to sprint. Dots just
/// say where you are.
struct StepDots: View {
    let current: Int
    let total: Int

    var body: some View {
        HStack(spacing: 6) {
            ForEach(0..<max(total, 1), id: \.self) { index in
                Circle()
                    .fill(index < current ? FocusColor.gold : FocusColor.outline)
                    .frame(width: index == current - 1 ? 7 : 5,
                           height: index == current - 1 ? 7 : 5)
            }
        }
    }
}

struct StatTile: View {
    let value: String
    let label: String
    var accent: Bool = false

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.xs) {
            Text(value)
                .font(.focusHeadline)
                .foregroundStyle(accent ? FocusColor.gold : FocusColor.textOnInk)
            Text(label)
                .font(.focusBody)
                .foregroundStyle(FocusColor.textMuted)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

File: app/src/main/java/com/focusdhikr/ui/theme/Theme.kt
package com.focusdhikr.ui.theme

import android.app.Activity
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.SideEffect
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat

/**
 * A quiet palette.
 *
 * Warm near-black and parchment rather than pure black and white, one muted
 * gold accent, and nothing else. No progress bars racing to full, no confetti,
 * no streak flames. The visual language should feel closer to a well-set book
 * than to a game. See requirement 12.
 */
object FocusColors {
    val Ink = Color(0xFF12100E)
    val InkElevated = Color(0xFF1C1917)
    val InkSubtle = Color(0xFF2A2624)

    val Parchment = Color(0xFFFAF9F6)
    val ParchmentElevated = Color(0xFFF2F0EA)

    val TextOnInk = Color(0xFFE8E3D8)
    val TextOnInkMuted = Color(0xFFA8A199)
    val TextOnParchment = Color(0xFF1C1917)
    val TextOnParchmentMuted = Color(0xFF6B645C)

    /** The single accent. Used sparingly: remaining time, the active step, dhikr. */
    val Gold = Color(0xFFC9A227)
    val GoldSoft = Color(0xFFD9BE63)

    /** Never red-for-shame. This is the "you have reached your limit" tone. */
    val Clay = Color(0xFFB07156)
}

private val DarkScheme = darkColorScheme(
    primary = FocusColors.Gold,
    onPrimary = FocusColors.Ink,
    secondary = FocusColors.GoldSoft,
    onSecondary = FocusColors.Ink,
    background = FocusColors.Ink,
    onBackground = FocusColors.TextOnInk,
    surface = FocusColors.InkElevated,
    onSurface = FocusColors.TextOnInk,
    surfaceVariant = FocusColors.InkSubtle,
    onSurfaceVariant = FocusColors.TextOnInkMuted,
    outline = FocusColors.InkSubtle,
    error = FocusColors.Clay,
    onError = FocusColors.TextOnInk,
)

private val LightScheme = lightColorScheme(
    primary = FocusColors.Gold,
    onPrimary = FocusColors.Parchment,
    secondary = FocusColors.Clay,
    onSecondary = FocusColors.Parchment,
    background = FocusColors.Parchment,
    onBackground = FocusColors.TextOnParchment,
    surface = FocusColors.ParchmentElevated,
    onSurface = FocusColors.TextOnParchment,
    surfaceVariant = Color(0xFFE7E3DA),
    onSurfaceVariant = FocusColors.TextOnParchmentMuted,
    outline = Color(0xFFD8D3C8),
    error = FocusColors.Clay,
    onError = FocusColors.Parchment,
)

/**
 * Serif for anything meant to be read slowly, sans for anything meant to be
 * scanned. Both are system families, so the app ships no font binaries.
 */
private val FocusTypography = Typography(
    displaySmall = TextStyle(
        fontFamily = FontFamily.Serif,
        fontWeight = FontWeight.Normal,
        fontSize = 34.sp,
        lineHeight = 42.sp,
    ),
    headlineMedium = TextStyle(
        fontFamily = FontFamily.Serif,
        fontWeight = FontWeight.Normal,
        fontSize = 26.sp,
        lineHeight = 34.sp,
    ),
    headlineSmall = TextStyle(
        fontFamily = FontFamily.Serif,
        fontWeight = FontWeight.Normal,
        fontSize = 21.sp,
        lineHeight = 29.sp,
    ),
    titleMedium = TextStyle(
        fontFamily = FontFamily.SansSerif,
        fontWeight = FontWeight.Medium,
        fontSize = 16.sp,
        lineHeight = 22.sp,
    ),
    bodyLarge = TextStyle(
        fontFamily = FontFamily.SansSerif,
        fontWeight = FontWeight.Normal,
        fontSize = 16.sp,
        lineHeight = 25.sp,
    ),
    bodyMedium = TextStyle(
        fontFamily = FontFamily.SansSerif,
        fontWeight = FontWeight.Normal,
        fontSize = 14.sp,
        lineHeight = 21.sp,
    ),
    labelLarge = TextStyle(
        fontFamily = FontFamily.SansSerif,
        fontWeight = FontWeight.Medium,
        fontSize = 14.sp,
        letterSpacing = 0.4.sp,
    ),
    labelSmall = TextStyle(
        fontFamily = FontFamily.SansSerif,
        fontWeight = FontWeight.Medium,
        fontSize = 11.sp,
        letterSpacing = 1.2.sp,
    ),
)

/** Spacing scale. Generous by default; whitespace is most of the calm. */
object Spacing {
    val xs = 4.dp
    val sm = 8.dp
    val md = 16.dp
    val lg = 24.dp
    val xl = 32.dp
    val xxl = 48.dp
}

@Composable
fun FocusDhikrTheme(
    darkTheme: Boolean = isSystemInDarkTheme(),
    /** The gate is always dark: it appears over other apps, often at night. */
    forceDark: Boolean = false,
    content: @Composable () -> Unit,
) {
    val useDark = forceDark || darkTheme
    val scheme = if (useDark) DarkScheme else LightScheme
    val view = androidx.compose.ui.platform.LocalView.current

    if (!view.isInEditMode) {
        val context = LocalContext.current
        SideEffect {
            (context as? Activity)?.window?.let { window ->
                WindowCompat.getInsetsController(window, view).apply {
                    isAppearanceLightStatusBars = !useDark
                    isAppearanceLightNavigationBars = !useDark
                }
            }
        }
    }

    MaterialTheme(
        colorScheme = scheme,
        typography = FocusTypography,
        content = content,
    )
}
